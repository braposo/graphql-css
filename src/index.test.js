/* eslint-env jest */

import React from "react";
import { render, Simulate } from "react-testing-library";
import serializer from "jest-glamor-react";
import { Div } from "glamorous";
import PropTypes from "prop-types";
import gqlCSS, { GqlCSS, GqlCSSProvider, WithGqlCSS, withGqlCSS } from "./index";
import styles from "../examples/styleguide";
import { h1Styles, h2Styles, customH1Styles, stateStyles } from "../examples/styleQueries";

expect.addSnapshotSerializer(serializer);

describe("gqlCSS", () => {
    it("it allows the extraction of styles by passing false", () => {
        const extractedStyles = gqlCSS(styles)(h2Styles, false);
        expect(extractedStyles).toEqual({
            color: "green",
            fontSize: "36px",
            fontWeight: 700,
            marginLeft: 32,
        });
    });

    it("it returns a styled component by passing a div", () => {
        const StyledComponent = gqlCSS(styles)(h2Styles, "div");
        const { container } = render(<StyledComponent />);
        expect(container).toMatchSnapshot();
    });

    it("it returns a styled component by not passing any component", () => {
        const StyledComponent = gqlCSS(styles)(h2Styles);
        const { container } = render(<StyledComponent />);
        expect(container).toMatchSnapshot();
    });

    it("it supports variables and stateful components", () => {
        class StatefulComponent extends React.Component {
            constructor(props) {
                super(props);
                this.state = {
                    variant: "normal",
                };
                this.handleClick = this.handleClick.bind(this);
            }

            handleClick() {
                this.setState(state => {
                    return {
                        variant: state.variant === "normal" ? "done" : "normal",
                    };
                });
            }

            render() {
                const { variant } = this.state;
                const OtherComponent = gqlCSS(styles)(stateStyles, null, { variant });

                return (
                    <React.Fragment>
                        <GqlCSS
                            styles={styles}
                            query={stateStyles}
                            variables={{ variant }}
                            onClick={this.handleClick}
                            data-testid="stateful-component"
                        >
                            Using stateful component
                        </GqlCSS>
                        <OtherComponent>Other sharing state</OtherComponent>
                    </React.Fragment>
                );
            }
        }

        const { container, queryByTestId } = render(<StatefulComponent />);

        expect(container).toMatchSnapshot();

        Simulate.click(queryByTestId("stateful-component"));

        expect(container).toMatchSnapshot();
    });
});

describe("GqlCSS", () => {
    it("it renders component without styles", () => {
        const { container } = render(<GqlCSS query={h2Styles}>Using component</GqlCSS>);

        expect(container).toMatchSnapshot();
    });
});

describe("GqlCSSProvider", () => {
    it("it supports the Provider/Subscriber pattern", () => {
        const SubscriberComponent = () => (
            <div>
                <div>
                    <span>
                        <GqlCSS query={h2Styles}>Using provider</GqlCSS>
                    </span>
                </div>
                <GqlCSS query={h1Styles} css={{ marginTop: 30 }}>
                    Also using provider
                </GqlCSS>
            </div>
        );
        const { container } = render(
            <GqlCSSProvider styles={styles}>
                <SubscriberComponent />
            </GqlCSSProvider>
        );

        expect(container).toMatchSnapshot();
    });
});

describe("WithGqlCSS", () => {
    it("it supports the render props pattern", () => {
        const { container } = render(
            <WithGqlCSS styles={styles} query={h2Styles}>
                {({ gqlStyles }) => <div style={gqlStyles}>Render props component</div>}
            </WithGqlCSS>
        );

        expect(container).toMatchSnapshot();
    });
});

describe("withGqlCSS", () => {
    it("it supports the HOC pattern", () => {
        const ExistingComponent = ({ gqlStyles, ...rest }) => (
            <Div css={gqlStyles} {...rest}>
                test
            </Div>
        );

        ExistingComponent.propTypes = {
            gqlStyles: PropTypes.object,
        };

        const HoC = withGqlCSS(styles, customH1Styles)(ExistingComponent);
        const { container } = render(<HoC />);

        expect(container).toMatchSnapshot();
    });
});
