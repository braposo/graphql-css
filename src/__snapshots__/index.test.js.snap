// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GqlCSS it renders component with styles 1`] = `
<div>
  <div
    class="x0 x1 x2 x3"
  >
    Using component with styles
  </div>
</div>
`;

exports[`GqlCSS it renders component without styles 1`] = `
<div>
  <div
    class=""
  >
    Using component without styles
  </div>
</div>
`;

exports[`useGqlCSS with GqlCSS it supports variables and stateful components 1`] = `
<div>
  <div
    class="x4 x5 x6 x7"
    data-testid="stateful-component"
  >
    Using stateful component
  </div>
</div>
`;

exports[`useGqlCSS with GqlCSS it supports variables and stateful components 2`] = `
<div>
  <div
    class="x4 x9 x6 x7"
    data-testid="stateful-component"
  >
    Using stateful component
  </div>
</div>
`;

exports[`useGqlCSS with styled it fails if interpolation is null 1`] = `
<div>
  <div
    class=""
  >
    Some heading
  </div>
</div>
`;

exports[`useGqlCSS with styled it fails if props don't exist 1`] = `
<div>
  <div
    class=""
  >
    Some heading
  </div>
</div>
`;

exports[`useGqlCSS with styled it handles style interpolation 1`] = `
<div>
  <h3
    class="x8"
  >
    Some heading
  </h3>
</div>
`;

exports[`useGqlCSS with styled it returns a styled component 1`] = `
<div>
  <h2
    class="x0 x1 x2 x3"
  >
    Some heading
  </h2>
</div>
`;

exports[`useGqlCSS with styled it supports variables and stateful components 1`] = `
<div>
  <div
    class="x4 x5 x6 x7 x8"
  >
    Other component sharing state
  </div>
</div>
`;

exports[`useGqlCSS with styled it supports variables and stateful components 2`] = `
<div>
  <div
    class="x4 x9 x6 x7 x8"
  >
    Other component sharing state
  </div>
</div>
`;
